lib/python/rose/macro.py

```diff
@@ -1276,23 +1298,92 @@ def _report_error(exception=None, text=""):
     )
 
 
+def get_apps_from_dir(conf_dir):
+    """Returns a list of apps with respect to the provided directory.
+    If the directory is an app return that app.
+    If the directory contains apps return the paths of those apps.
+    If the directory is a suite return the paths of its apps."""
+    config_file_path = os.path.join(conf_dir,
+                                    rose.SUB_CONFIG_NAME)
+    if (os.path.exists(config_file_path) or os.path.isfile(config_file_path)):
+        # we are in an app
+        return DIR_TYPE_APP, [conf_dir]
+    else:
+        # are we in an apps folder?
+        apps = []
+        for _, subdirs, _ in os.walk(conf_dir):
```

Comment:
Sorry, we could do with doing `glob.glob` here as well - or actually, it might be better to do what `rose edit` does and just go down the directory tree until we find a relevant configuration file. That avoids the problem of being stuck in e.g. `etc/foo/bar/` suite subdirectory and running `rose macro`. `rose edit` in `rose/config_editor/main.py` does:

```python
    path = os.getcwd()
    name_set = set([rose.SUB_CONFIG_NAME, rose.TOP_CONFIG_NAME])
    while True:
        if set(os.listdir(path)) & name_set:
            break
        path = os.path.dirname(path)
        if path == os.path.dirname(path):
            # We don't support suites located at the root!
            break
    if path != os.getcwd() and path != os.path.dirname(path):
        os.chdir(path)
```

Maybe we could pull this out into a new function and call that?